We’re still sometimes experiencing hanging tests that pytest can’t kill, even with this change. But it *does* enable me to run all tests on my Mac. I also observe these hangs locally; aborting them reveals which tests are affected. This PR doesn't resolve all issues, and I haven’t encountered the errors reported in the issue.

Tests running for over 2 minutes were mostly not automatically finished by pytest; I had to manually kill them with Ctrl+C.

```
=========== 576 passed, 16 skipped, 92 warnings in 105.15s (0:01:45) ===========
=========== 576 passed, 16 skipped, 92 warnings in 100.00s (0:01:39) ===========
=========== 576 passed, 16 skipped, 92 warnings in 102.45s (0:01:42) ===========
====== 1 failed, 575 passed, 16 skipped, 92 warnings in 198.85s (0:03:18) ======
====== 2 failed, 574 passed, 16 skipped, 92 warnings in 169.84s (0:02:49) ======
====== 1 failed, 575 passed, 16 skipped, 92 warnings in 973.18s (0:16:13) ======
=========== 576 passed, 16 skipped, 92 warnings in 102.40s (0:01:42) ===========
=========== 576 passed, 16 skipped, 92 warnings in 102.92s (0:01:42) ===========
=========== 576 passed, 16 skipped, 92 warnings in 106.08s (0:01:46) ===========
====== 2 failed, 574 passed, 16 skipped, 92 warnings in 389.04s (0:06:29) ======
=========== 576 passed, 16 skipped, 92 warnings in 102.96s (0:01:42) ===========
=========== 576 passed, 16 skipped, 92 warnings in 102.52s (0:01:42) ===========
=========== 576 passed, 16 skipped, 92 warnings in 107.22s (0:01:47) ===========
=========== 576 passed, 16 skipped, 92 warnings in 105.72s (0:01:45) ===========
====== 1 failed, 575 passed, 16 skipped, 92 warnings in 102.76s (0:01:42) ======
=========== 576 passed, 16 skipped, 92 warnings in 101.19s (0:01:41) ===========
=========== 576 passed, 16 skipped, 92 warnings in 103.57s (0:01:43) ===========
=========== 576 passed, 16 skipped, 92 warnings in 105.69s (0:01:45) ===========
=========== 576 passed, 16 skipped, 92 warnings in 100.17s (0:01:40) ===========
=========== 576 passed, 16 skipped, 92 warnings in 107.49s (0:01:47) ===========
=========== 576 passed, 16 skipped, 92 warnings in 133.30s (0:02:13) ===========
====== 2 failed, 574 passed, 16 skipped, 92 warnings in 414.41s (0:06:54) ======
```
If there is no [gw*] prefix, it indicates a test that has hung; all others have failed normally. This information wasn’t easily accessible before.

```

[gw11] [ 75%] FAILED tests/_test_msui/test_topview.py::Test_TopViewWMS::test_server_getmap
FAILED tests/_test_msui/test_topview.py::Test_TopViewWMS::test_server_getmap
[gw2] [ 82%] FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_server_service_cache
[gw2] [ 90%] FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_server_getmap
FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_server_service_cache
FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_server_getmap
[gw15] [ 82%] FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_multilayer_syncing
FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_multilayer_syncing
[gw13] [ 44%] FAILED tests/_test_msui/test_sideview.py::Test_SideViewWMS::test_server_getmap
[gw7] [ 88%] FAILED tests/_test_msui/test_wms_control.py::Test_VSecWMSControlWidget::test_server_getmap
FAILED tests/_test_msui/test_sideview.py::Test_SideViewWMS::test_server_getmap
FAILED tests/_test_msui/test_wms_control.py::Test_VSecWMSControlWidget::test_server_getmap
[gw7] [ 38%] FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_no_server
FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_no_server
[gw14] [ 93%] FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_no_server
[gw15] [ 95%] FAILED tests/_test_msui/test_wms_control.py::Test_VSecWMSControlWidget::test_server_getmap
FAILED tests/_test_msui/test_wms_control.py::Test_HSecWMSControlWidget::test_no_server
FAILED tests/_test_msui/test_wms_control.py::Test_VSecWMSControlWidget::test_server_getmap
```

For my Mac-based development, this is a big step forward, and I don't believe it will regress the Ubuntu test results.  Based on that, I think we should merge.

--- a/tests/fixtures.py
+++ b/tests/fixtures.py
@@ -28,7 +28,6 @@ import mock
 import multiprocessing
 import time
 import urllib
-import socket
 import socketio
 import mslib.mswms.mswms
 import eventlet
@@ -187,22 +186,13 @@ def mswms_server(mswms_app):
         yield url
 
 
-def _reserve_port(host):
-    """Only reserve the port for the server"""
-    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
-        # force IPv4 with AF_INET and SOCK_STREAM
-        s.bind((host, 0))
-        # make the port reusable immediately after closing the socket
-        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        port = s.getsockname()[1]
-        return port
-
-
-def _start_eventlet_server(host, port, app):
+def _start_eventlet_server(host, port_queue, app):
     """
-    Start the eventlet server for the given app on the specified host and port.
+    Starts the Eventlet server and sends the chosen port back to the parent process.
     """
-    sock = eventlet.listen((host, port))
+    sock = eventlet.listen((host, 0))
+    port = sock.getsockname()[1]
+    port_queue.put(port)
     eventlet.wsgi.server(sock, app, log_output=False)
 
 
@@ -211,15 +201,26 @@ def _running_eventlet_server(app):
     """Context manager that starts the app in an eventlet server and returns its URL."""
     scheme = "http"
     host = "127.0.0.1"
-    port = _reserve_port(host)
-    url = f"{scheme}://{host}:{port}"
-    app.config['URL'] = url
+
     if "fork" not in multiprocessing.get_all_start_methods():
         pytest.skip("requires the multiprocessing start_method 'fork', which is unavailable on this system")
+
     ctx = multiprocessing.get_context("fork")
-    process = ctx.Process(target=_start_eventlet_server, args=(host, port, app), daemon=True)
+    # We are using a queue to retrieve the port selected in the child process.
+    port_queue = ctx.Queue()
+
+    process = ctx.Process(target=_start_eventlet_server, args=(host, port_queue, app), daemon=True)
     try:
         process.start()
+        # Retrieve the port from the queue
+        try:
+            port = port_queue.get(timeout=10)
+        except multiprocessing.queues.Empty:
+            raise RuntimeError("Could not retrieve port from server process")
+
+        url = f"{scheme}://{host}:{port}"
+        app.config['URL'] = url
+
         start_time = time.time()
         sleep_time = 0.01
         time_out = 20
@@ -228,9 +229,7 @@ def _running_eventlet_server(app):
         while not is_url_response_ok(readiness_url):
             if not process.is_alive():
                 # show the exitcode for further debugging
-                raise RuntimeError(
-                    f"Server process exited early with code {process.exitcode} at {url}"
-                )
+                raise RuntimeError(f"Server process exited early with code {process.exitcode} at {url}")
             if (time.time() - start_time) > time_out:
                 raise RuntimeError(f"Server did not start within {time_out} seconds at {url}")
             time.sleep(sleep_time)

--- a/tests/fixtures.py
+++ b/tests/fixtures.py
@@ -192,8 +192,9 @@ def _reserve_port(host):
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
         # force IPv4 with AF_INET and SOCK_STREAM
         s.bind((host, 0))
+        # make the port reusable immediately after closing the socket
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         port = s.getsockname()[1]
-        s.close()
         return port
 
 

